### PR TITLE
Set request context in GridBase

### PR DESCRIPTION
**Pull request by bchopson**
- GridBase gets request context from grid manager
- Indicate inrequest is internal to Webgrid

Fixes GH-141

### CODE REVIEW

File: webgrid/testing.py [hunk 1]
```diff
@@ -277,12 +277,29 @@ class GridBase:
...
     def setup_class(cls):
         if hasattr(cls, 'init'):
             cls.init()
 
+    @property
+    def url(self):
+        return self._url
+
+    @contextmanager
+    def url_context(self, url):
```
Comment (bchopson) [inline webgrid/testing.py:292]: I'm open to other approaches if this context manager seems a bit weird. It seemed like the best way to ensure that all calls to get_session_grid used the same URL for the request context, without explicitly passing the url to each dependent method.

File: webgrid_ta/helpers.py [hunk 2]
```diff
@@ -71,13 +66,3 @@ def assert_not_in_query(obj, test_for):
         query = obj
     query_str = query_to_str(query)
     assert test_for not in query_str, query_str
-
-
-def inrequest(*req_args, **req_kwargs):
```
Comment (bchopson) [inline webgrid_ta/helpers.py:76]: dead code

File: webgrid/extensions.py [hunk 3]
```diff
@@ -418,3 +418,7 @@ def get_args(self, grid):
             args = loader(self).get_args(grid, args)
 
         return args
+
+    def request_context(self, url='/'):
```
Comment (guruofgentoo) [inline webgrid/extensions.py:422]: The docstring specifies tests, but we should call the method itself `test_request_context`.

File: webgrid/flask.py [hunk 4]
```diff
@@ -1,6 +1,14 @@
 from __future__ import absolute_import
 
-from flask import request, session, flash, Blueprint, url_for, send_file
+from flask import (
```
Comment (guruofgentoo) [inline webgrid/flask.py:3]: nit: we probably should have just done `import flask` to begin with. Oh well. No real need to change it right now.

File: webgrid/testing.py [hunk 5]
```diff
@@ -343,8 +360,9 @@ def get_session_grid(self, *args, **kwargs):
             grid instance
         """
         grid = self.grid_cls(*args, **kwargs)
-        grid.apply_qs_args()
-        return grid
+        with grid.manager.request_context(url=self.url):
```
Comment (guruofgentoo) [inline webgrid/testing.py:363]: If the surrounding test has set up a request context already, we should preserve that here. Need to check the stack to see if one has been pushed.

File: webgrid/testing.py (hunk 1 above)
Comment (guruofgentoo) [reply webgrid/testing.py:292]: I'd rather not - we're trying to shortcut the need to use a context manager to call `get_session_grid`. Instead, let's have a special kwarg to pass to `get_session_grid`, like `_query_string`, and just pass that into the request context.

**Review by guruofgentoo — CHANGES_REQUESTED** [review]
Also needs a `get_grid` method on the base that takes args to supply directly to `apply_qs_args`. That won't need the request context in most cases, which will reduce some of the overhead.

File: webgrid/flask.py (hunk 4 above)
Comment (bchopson) [reply webgrid/flask.py:3]: I'll go ahead and do this, as it is a major improvement in readability.

**Pull request merged**